test/jdk/java/awt/dnd/DnDClipboardDeadlockTest.java

```diff
@@ -0,0 +1,441 @@
+        robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+        synchronized (SYNC_LOCK) {
+            robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+            SYNC_LOCK.wait(MOUSE_RELEASE_TIMEOUT);
+        }
+
+        Component c = clickedComponent;
+
+        while (c != null && c != comp) {
+            c = c.getParent();
+        }
+
+        return c == comp;
+    }
+
+    public static void waitForInit() throws InterruptedException {
+        final Frame f = new Frame() {
+            public void paint(Graphics g) {
+                dispose();
+                synchronized (SYNC_LOCK) {
+                    SYNC_LOCK.notifyAll();
+                }
+            }
+        };
+        f.setBounds(600, 400, 200, 200);
+        synchronized (SYNC_LOCK) {
+            f.setVisible(true);
+            SYNC_LOCK.wait(PAINT_TIMEOUT);
+        }
+        tk.sync();
+    }
+}
+
+class DragSourceButton extends Button implements Serializable,
+        DragGestureListener,
+        DragSourceListener {
+    static final Clipboard systemClipboard =
+            Toolkit.getDefaultToolkit().getSystemClipboard();
+    final Transferable transferable;
+
+    public DragSourceButton(Transferable t) {
+        super("DragSourceButton");
+
+        this.transferable = t;
+        DragSource ds = DragSource.getDefaultDragSource();
+        ds.createDefaultDragGestureRecognizer(this, DnDConstants.ACTION_COPY,
+                this);
+    }
+
+    public void dragGestureRecognized(DragGestureEvent dge) {
+        dge.startDrag(null, transferable, this);
+    }
+
+    public void dragEnter(DragSourceDragEvent dsde) {
+    }
+
+    public void dragExit(DragSourceEvent dse) {
+    }
+
+    public void dragOver(DragSourceDragEvent dsde) {
+        try {
+            Transferable t = systemClipboard.getContents(null);
+            if (t.isDataFlavorSupported(DataFlavor.stringFlavor)) {
+                String str = (String) t.getTransferData(DataFlavor.stringFlavor);
+            }
+            systemClipboard.setContents(new StringSelection("SOURCE"), null);
+        } catch (IOException ioe) {
+            ioe.printStackTrace();
+            if (!ioe.getMessage().equals("Owner failed to convert data")) {
+                throw new RuntimeException("Owner failed to convert data");
+            }
+        } catch (IllegalStateException e) {
+            // IllegalStateExceptions do not indicate a bug in this case.
+            // They result from concurrent modification of system clipboard
+            // contents by the parent and child processes.
+            // These exceptions are numerous, so we avoid dumping their
+            // backtraces to prevent blocking child process io, which
+            // causes test failure on timeout.
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void dragDropEnd(DragSourceDropEvent dsde) {
+        System.exit(DnDClipboardDeadlockTest.CODE_OK);
```

Comment:
SFAICT you haven't done anything to ensure this is ONLY the child process calling System.exit()

Reply:
Do you mean via a lock or am I misunderstanding this? I assumed this `exit()` was called once because there is only one instance of DragSourceButton/DragSourcePanel which would be the only one calling `exit()`.

Reply:
> Do you mean via a lock or am I misunderstanding this? I assumed this `exit()` was called once because there is only one instance of DragSourceButton/DragSourcePanel which would be the only one calling `exit()`.

OK. If that's the case then this should be fine

Reply:
The above explanation sounds correct. System.exit() is called only from child process.

Parent process : test.start() creates DropTargetPanel()
Child process: test.run(args) creates DragSourcePanel + DragSourceButton => DragSourceButton has the exit() call.